### PR TITLE
RandomForest interface

### PR DESCRIPTION
**Pull request by luizfgoncalves**
Can @raphaelcampos, @clebsonc , @thiagovas and @gcmgomes review it?
I included everyone because we were discussing about making it an abstract class (like in the tree branch).

### CODE REVIEW

File: ensemble/randomforest.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#ifndef _MACHINELEARNING_ENSEMBLE_RANDOMFOREST_H_
+#define _MACHINELEARNING_ENSEMBLE_RANDOMFOREST_H_
+
+#include "../tree/tree.h"
```
Comment (gcmgomes) [inline ensemble/randomforest.h]: Use absolute include paths. In this case just `tree/tree.h` will do the trick. Bazel takes care of them for use

File: ensemble/randomforest.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#ifndef _MACHINELEARNING_ENSEMBLE_RANDOMFOREST_H_
```
Comment (gcmgomes) [inline ensemble/randomforest.h]: Change the file name to `random_forest.h` and the header guard to reflect it.

File: ensemble/randomforest.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+   * all possible classes. They are the mean predicted class
+   * probabilities of the trees in the forest
+   * \param Instance probability distribution
+   */
+  virtual void PredictProbability(/* Collection col, */
+                                  std::vector<std::vector<double> >
+                                      *instance_probability) = 0;
+};
+
+} //end namespace ensemble
```
Comment (gcmgomes) [inline ensemble/randomforest.h]: The end is pointless here, just drop it.

File: ensemble/randomforest.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ */
+class RandomForest {
+ public:
+  /**
+   * \brief Standard destructor.
+   */
+  virtual ~RandomForest();
+
+  /**
+   * Compute and return the score of the training dataset obtained using an
```
Comment (gcmgomes) [inline ensemble/randomforest.h]: Change this to something like `Computes the out-of-bag score of the training set`.

File: ensemble/randomforest.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+   * out-of-bag estimate
+   * \return oob score
+   */
+  virtual double OobScore() = 0;
+
+  /*
+   * Build the RandomForest
```
Comment (gcmgomes) [inline ensemble/randomforest.h]: Builds instead of build, and add a final dot.

File: ensemble/randomforest.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+   * \brief Standard destructor.
+   */
+  virtual ~RandomForest();
+
+  /**
+   * Compute and return the score of the training dataset obtained using an
+   * out-of-bag estimate
+   * \return oob score
+   */
+  virtual double OobScore() = 0;
```
Comment (clebsonc) [inline ensemble/randomforest.h]: Is this function returning the score for each sample in the Oob or just the coefficient of determination?

Since it is returning just a double value I am assuming that it is returning a metric value that represents the prediction for the entire set of samples in the Oob.

If this is the case, I think that we should also provide a function that returns the prediction for each sample in the Oob.

What do you guys think? @raphaelcampos @thiagovas @gcmgomes

File: ensemble/randomforest.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  /*
+   * Build the RandomForest
+   * \param thread Number of threads for paralel execution
+   * \param tree Number of trees in the Forest
+   */
+  virtual void Build(/* Collection col, */ const int thread,
+                     const int tree) = 0;
+
+  /**
+   * Predicts the labels of the instances of X based on a vote by
```
Comment (clebsonc) [inline ensemble/randomforest.h]: I think that for this function we should provide the brief and a larger comment explaining how this prediction is made. Also, since this is an interface, not all random forests are based on the vote (which I am assuming here that you meant majority vote). Please Provide a better explanation of the tasks that this function will do.

What do you guys think? @raphaelcampos @thiagovas @gcmgomes

File: ensemble/randomforest.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+   * the trees of the forest, weighted by their probability estimates
+   * \param Instance labels
```
Comment (clebsonc) [inline ensemble/randomforest.h]: If this function is predicting the class in the context of classification of the regression score for regression problems, why it is required to pass the labels of the samples?
I think that it is not making that much sense. If you are trying to predict something, just pass the collection of samples that will be predicted, and then return a vector containing its prediction.

What do you guys think? @raphaelcampos @thiagovas @gcmgomes

File: ensemble/randomforest.h [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+   */
+  virtual void Predict(/* Collection col, */
```
Comment (clebsonc) [inline ensemble/randomforest.h]: As I mentioned above, we should return a vector containing the prediction of each sample in the training set.

What do you guys think? @raphaelcampos @thiagovas @gcmgomes

File: ensemble/randomforest.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+                       std::vector<int> *instance_label) = 0;
+
+  /**
+   * Returns the predicted probability distribution of each instance in X over
+   * all possible classes. They are the mean predicted class
+   * probabilities of the trees in the forest
+   * \param Instance probability distribution
+   */
+  virtual void PredictProbability(/* Collection col, */
```
Comment (clebsonc) [inline ensemble/randomforest.h]: I still do not get what this function is computing. What are the differences of this one from the `Predict`?
Provide a better explanation in the comments and a better name for the previously one that better describes it.

What do you guys think? @raphaelcampos @thiagovas @gcmgomes

File: ensemble/randomforest.h (hunk 9 above)
Comment (gcmgomes) [inline ensemble/randomforest.h]: I'd actually implement the prediction of a single sample before we worry about multi-sample predictions. It feels more natural and will eventually be used on multi-sample prediction.

File: ensemble/randomforest.h (hunk 8 above)
Comment (gcmgomes) [inline ensemble/randomforest.h]: Agreed. This is starting to feel more like a task for the `Classifier` and `Regressor` interfaces we will be implementing later on.

File: ensemble/randomforest.h (hunk 6 above)
Comment (gcmgomes) [inline ensemble/randomforest.h]: Out of bag is also a measure dependent upon the type of model we are training, i.e classifier/regressor.

File: ensemble/randomforest.h (hunk 9 above)
Comment (clebsonc) [inline ensemble/randomforest.h]: Ok, so why receive a collection and return only the prediction for a single sample? Doesn't make much sense for me.
We can see so far that the interface was receiving a possible collection and its labels.
So why do not return the prediction of this collection?

File: ensemble/randomforest.h (hunk 7 above)
Comment (gcmgomes) [inline ensemble/randomforest.h]: As it turns out, it looks like all of our algorithms will effectively only handle the `Build` or `Train` methods themselves, and then their "usable" versions will need to implement the `Predict` or `Score`  inherited from the appropriate model type.

**Comment by luizfgoncalves** [comment]
I fixed the minor problems and included a function called OobScoreSample as requested by @clebsonc. 

About the "Predict" function, what I thought was:
Given the test set (Collection col) it'll return the labels for each instance. As we discussed yesterday, if I want to return a vector it should be a pointer in the parameters of the function, and that is what instance_label represents.

Right now I feel it'll only work with classification (as it is of type int). Should I change for a genenic one of type double, so it works with regression and classification?

PredictProbability is specific to classification too (as @raphaelcampos mentioned on Hangouts), should I just drop it on this interface?

**Comment by raphaelcampos** [comment]
Luiz, look at the pull request #24. The prediction signature belongs to the Learner interface, alongside with the methods PredictProbability and Train. The idea we were discussing on friday was that the useful models (Regressor/Classifier) will implement the learner interface with their own needs and also inherit from model (e.g. RandomForest) which implements the common methods shared by the Regressor and Classifier.

File: ensemble/random_forest.h [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+   */
+  virtual double OobScore() = 0;
+
+  /**
+   * Compute the out-of-bag score that represents the prediction 
+   * for each sample in the oob.
+   * \param tree_idx Index of the tree
+   * \return oob score
+   */
+  virtual double OobScoreSample(const int tree_idx) = 0;
```
Comment (clebsonc) [inline ensemble/random_forest.h]: Since we are returning the prediction of each sample in the out-of-bog, why the return value here is just a single double? Shouldn't be a vector or something like that, in which contains the index of the document (this is the document in the oob) and its score?

**Comment by luizfgoncalves** [comment]
RandomForest is now a concrete class that extends Learner. I've included a constructor and the variables trees_ and threads_. 

Made some changes to the OOB functions too, according to what @clebsonc said. They are commented right now, because an important discussion is if they should be members of the RandomForest class, as OOB is a concept present in many models utilizing bootstrap agregation (not exclusively RFs).

**Comment by luizfgoncalves** [comment]
What do you guys think about the new header file?

File: ensemble/random_forest.h [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    ~RandomForest();
+
+    /**
+     * Train the learner with a given dataset.
+     * \param X A pointer object of Collection.
+     */  
+    void Train(/*const Collection *X*/);
+
+    /**
```
Comment (gcmgomes) [inline ensemble/random_forest.h]: Lets handle OOB in a later pull request. Remove these comments and we will be good to go.

File: ensemble/random_forest.h [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    RandomForest(int tree_count, int thread_count);
+
+    /**
+     * \brief Standard destructor.
+     */
+    ~RandomForest();
+
+    /**
+     * Train the learner with a given dataset.
```
Comment (gcmgomes) [inline ensemble/random_forest.h]: Change Train in the comment for Trains.

**Comment by luizfgoncalves** [comment]
Fixed. Should I work on the .cpp file for this class on this pull request too?

**Comment by gcmgomes** [comment]
No, do it in a future PR. LGTM, approval.
